packages/flutter/lib/src/widgets/interactive_viewer.dart

```diff
@@ -1272,3 +1205,298 @@ Axis? _getPanAxis(Offset point1, Offset point2) {
+    }
+
+    // If the slopes are different and not vertical, the corresponding lines
+    // (not segments) must overlap. If that point happens between the endpoints
+    // of both line segments, then the line segmens intersect.
+    // x = (b1 - b) / (m - m1)
+    final double intersectionX = (lineSegment.lineYIntercept - lineYIntercept)
+        / (slope - lineSegment.slope);
+    assert(intersectionX.isFinite);
+
+    // y = m * intersectionX + b
+    final double intersectionY = slope * intersectionX + lineYIntercept;
+    assert(intersectionY.isFinite);
+    final Offset intersection = Offset(intersectionX, intersectionY);
+
+    return contains(intersection) && lineSegment.contains(intersection);
+  }
+
+  /// Returns true iff this line segment intersects the given rect, inclusively.
+  bool intersectsRect(Rect rect) {
+    // Intersects if either point, or both, is inside of the rectangle,
+    // inclusively.
+    if (_rectContainsInclusive(rect, p0) || _rectContainsInclusive(rect, p1)) {
+      return true;
+    }
+
+    // Otherwise, intersects the rect if it intersects any of the four sides.
+    final LineSegment top = LineSegment(rect.topLeft, rect.topRight);
+    final LineSegment right = LineSegment(rect.topRight, rect.bottomRight);
+    final LineSegment bottom = LineSegment(rect.bottomRight, rect.bottomLeft);
+    final LineSegment left = LineSegment(rect.bottomLeft, rect.topLeft);
+    return intersects(top) || intersects(right)
+        || intersects(bottom) || intersects(left);
+  }
+
+  /// Assuming that the given point is on the extended line of this line segment,
+  /// returns the closest point on the line segment to it.
+  @visibleForTesting
+  Offset findClosestOffsetOnLineSegmentToOffsetOnLine(Offset pointOnLine) {
+    assert(_lineContains(pointOnLine));
+    if (contains(pointOnLine)) {
+      return pointOnLine;
+    }
+    final double d0 = _distanceBetweenPoints(pointOnLine, p0);
+    final double d1 = _distanceBetweenPoints(pointOnLine, p1);
+    return d0.abs() <= d1.abs() ? p0 : p1;
+  }
+
+  /// Returns the closest point on each [LineSegment] to the other
+  /// [LineSegment].
+  ///
+  /// In the returned [ClosestPoints], [ClosestPoints.a] represents the point on
+  /// this [LineSegment], and [ClosestPoints.b] represents the point on the
+  /// given [LineSegment].
+  @visibleForTesting
+  ClosestPoints findClosestPointsLineSegment(LineSegment lineSegment) {
+    final Offset? lineIntersection = linesIntersectAt(lineSegment);
+
+    // If the lines don't intersect (the line segments are parallel), then there
+    // are many closest points. This will return p0 and the closest point to p0.
+    if (lineIntersection == null) {
+      final Offset closestToP0 = lineSegment.findClosestToOffset(p0);
+      return ClosestPoints(
+        a: p0,
+        b: closestToP0,
+      );
+    }
+
+    // Otherwise, the closest points on the line segments are the closest points
```

Comment:
You're right, thanks for catching this!  I feel like the behavior near the boundary is a bit smoother after fixing this even. 
 I had a bunch of tests that were slightly off because of this too.